### DISCUSSION TITLE
Issue with Running Codemods for Migrating to React Query v5 Due to ESM Compatibility

### DISCUSSION THREAD

**Discussion by nadavschwartz** [TanStack/query#7048] [Q&A] [answered]
I've encountered a problem when trying to run the provided codemods for migrating to React Query v5, specifically related to the ESM compatibility issues in Node.js. I followed the migration guide and used jscodeshift to apply the remove-overloads.js codemod, but the process fails due to ESM module loading constraints.

### Summary

I've encountered a problem when trying to run the provided codemods for migrating to React Query v5 (from v4), specifically related to the "ESM compatibility issues in Node.js". I followed the migration guide and used `jscodeshift` to apply the `remove-overloads.js` codemod, but the process fails due to ESM module loading constraints.

### Environment

- **Node.js Version:** 18.18.2
- **React Query Version:** v5
- **Operating System:** macOS 14.1.1

### Steps I followed

1. Installed React Query v5 following the [official migration guide](https://tanstack.com/query/v5/docs/framework/react/guides/migrating-to-v5).
2. From the root of my application, attempted to run the `jscodeshift` tool with the provided codemod script for removing method overloads. I used `"."` in the command to specify the current directory as the target for the transformation, indicating I'm running it from the root application path:
   ```bash
   npx jscodeshift@latest .  \
     --extensions=ts,tsx \
     --parser=tsx \
     --transform=./node_modules/@tanstack/react-query/build/codemods/src/v5/remove-overloads/remove-overloads.js

Encountered the following error indicating an issue with ESM compatibility:
 
 `Error [ERR_REQUIRE_ESM]: require() of ES Module /path/to/node_modules/@tanstack/react-query/build/codemods/src/v5/remove-overloads/remove-overloads.js from /path/to/npm/_npx/a3dac9416cb5dcf1/node_modules/jscodeshift/src/Worker.js not supported.
remove-overloads.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules. Instead rename remove-overloads.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /path/to/node_modules/@tanstack/react-query/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).`

the output at the end of the script is as follows:

```bash
Node.js v18.18.2
All done. 
Results: 
0 errors
0 unmodified
0 skipped
0 ok
Time elapsed: 0.320seconds

**Comment by TkDodo** [comment]
@lachlancollins does tsup add `type:module` to the packages we ship, because in the checked-in package.json files, it's not there. If that's the case - do we need to rename the codemod files to `.cjs` ?

@nadavschwartz can you try to just rename the file to `remove-overloads.cjs` in your node_modules and see if it works?

**Reply by lachlancollins** [reply]
Oh yeah, I'm seeing now that they are CJS files. The nearest `package.json` is the root-level one for react-query which has `"type": "module"`. They should have the `.cjs` extension.

**Reply by TkDodo** [reply]
If renaming "just works", someone please PR it and update the instructions :)

**Reply by nadavschwartz** [reply]
Unfortunately, it didn't work :/
I renamed `remove-overloads.js` to `remove-overloads.cjs` and updated the script to be as follows:
```
npx jscodeshift@latest .  \
  --extensions=ts,tsx \
  --parser=tsx \
  --transform=./node_modules/@tanstack/react-query/build/codemods/src/v5/remove-overloads/remove-overloads.cjs
  ```
  
  when I ran the code above, it gave me the same error I got originally, about the following file:
  `..../node_modules/@tanstack/react-query/build/codemods/src/v5/remove-overloads/transformers/filter-aware-usage-transformer.js`
  
  so I updated the above file to be .cjs as well, ran the script, and got the  same error I got originally about the other file in the same  `transformers` folder:
  `....node_modules/@tanstack/react-query/build/codemods/src/v5/remove-overloads/transformers/query-fn-aware-usage-transformer.js`
  
 So I updated the above file too to end with `.cjs`, ran the script, and got this error:
 
 ```Error: Package subpath './build/codemods/src/utils/index.cjs' is not defined by "exports" in ..../node_modules/@tanstack/react-query/package.json imported from ..../node_modules/@tanstack/react-query/build/codemods/src/v5/remove-overloads/remove-overloads.cjs```
 
 so I opened that file and added that subpath
 
 before:
 ```
   "exports": {
    ".": {
      "import": {
        "types": "./build/modern/index.d.ts",
        "default": "./build/modern/index.js"
      },
      "require": {
        "types": "./build/modern/index.d.cts",
        "default": "./build/modern/index.cjs"
      }
    },
    "./package.json": "./package.json",
  },
  ```
  
  after:
  ```
    "exports": {
    ".": {
      "import": {
        "types": "./build/modern/index.d.ts",
        "default": "./build/modern/index.js"
      },
      "require": {
        "types": "./build/modern/index.d.cts",
        "default": "./build/modern/index.cjs"
      }
    },
    "./package.json": "./package.json",
    "./build/codemods/src/utils/index.cjs": "./build/codemods/src/utils/index.cjs"
  },
  ```
  
  
  I ran the script, and got this error:
  ```
  Error: Package subpath './build/codemods/src/v5/remove-overloads/transformers/filter-aware-usage-transformer' is not defined by "exports" in .../node_modules/@tanstack/react-query/package.json imported from ..../node_modules/@tanstack/react-query/build/codemods/src/v5/remove-overloads/remove-overloads.cjs
  ```
  
  so I added  `./build/codemods/src/v5/remove-overloads/transformers/filter-aware-usage-transformer` right below `"./build/codemods/src/utils/index.cjs"` in the `package.json`
  
  ran the script and got the following error about `package subpath` for the following:
  - `./build/codemods/src/v5/remove-overloads/utils/unknown-usage-error.js`
  -  `@tanstack/react-query/build/codemods/src/v5/remove-overloads/utils/index.js`
  
  and a few others. but at that point, I've lost track.  
  
  @TkDodo & @lachlancollins, I want to thank you both for responding.

**Reply by TkDodo** [reply]
maybe we need our own `package.json` then for the codemods?

**Reply by nadavschwartz** [reply]
Would it help if I shared with you my current `package.json`? perhaps it'll be of any benefit!
here it is: [link](https://gist.github.com/nadavschwartz/94caab271e639267d28933b3d2954b2c)

**Reply by smashercosmo** [reply]
One solution is to temporary change `type: "module"` to `type: "commonjs"` in react-query's package.json when running the script.

**Reply by ontanj** [reply, marked as answer]
the issue is resolved either by renaming all CommonJS files in codemods to .cjs, or by adding the package.json in the published codemods directory. since the project mainly uses ES modules it makes sense to name the CommonJS files .cjs. I created a pull request https://github.com/TanStack/query/pull/7333
